docs/settings-manager.md

```diff
@@ -0,0 +1,68 @@
+# Settings Manager
+
+Settings managers are used to persist configuration data like API keys, preferences, etc..
+
+## `SettingsValue` interface
+
+Entries in the manager are stored using structure:
+
+```
+  key: string;
+  value: string;
+  domain?: string;
+```
+
+For values which should not be migrated during environment cloning (as private keys to payment provider), developer should use domain field to bind it to particular store instance.
+
+## Available methods
+
+- `get: (key: string, domain?: string) => Promise<string | undefined>`
+- `set: (settings: SettingsValue[] | SettingsValue) => Promise<void>`
+
+# MetadataManager
+
+Default manager used by app template. Use app metadata as storage. Since app developer can use any GraphQL client, constructor must be parametrized with fetch and mutate functions:
+
+```ts
+import { MetadataEntry } from "@saleor/app-sdk/settings-manager";
+import { Client } from "urql";
+
+import {
+  FetchAppDetailsDocument,
+  FetchAppDetailsQuery,
+  UpdateAppMetadataDocument,
+} from "../generated/graphql";
+
+export async function fetchAllMetadata(client: Client): Promise<MetadataEntry[]> {
+  const { error, data } = await client
+    .query<FetchAppDetailsQuery>(FetchAppDetailsDocument, {})
+    .toPromise();
+
+  return data?.app?.privateMetadata.map((md) => ({ key: md.key, value: md.value })) || [];
+}
+
+export async function mutateMetadata(client: Client, metadata: MetadataEntry[]) {
+  const { error: mutationError, data: mutationData } = await client
+    .mutation(UpdateAppMetadataDocument, {
+      id: appId,
+      input: metadata,
+    })
+    .toPromise();
+
+  return (
+    mutationData?.updatePrivateMetadata?.item?.privateMetadata.map((md) => ({
+      key: md.key,
+      value: md.value,
+    })) || []
+  );
+}
```

Comment:
But it would require URQL as a dependency, which we wanted to avoid.

Maybe the example should be more minimal to not obfuscate the usage?

Reply:
It doesnt have to require URL. It can require an interface that urql passes (dependency inversion):
```
type Client = {
  mutation(...): {toPromise:() => ...}}
  query(...): {toPromise: () => ...}} 
```

Im not sure this is the best solution, but Im also not sure about leaking too much outside. We should try to hide Saleor stuff inside SDK if possible...

btw, maybe in this version of SDK we can actually include urql in SDK (or require in peer deps)? It can be unblocking, and we can always change it later. We are not sure if MetadataManager is something that will stay forever

Reply:
I would _really_ like to avoid depending on any gql client in the SDK, since it usually adds a bunch of dependencies and issues (graphql toolset seems to be picky with graphqljs/graphql-tag versions, at least that was my experience so far with such approaches).

If we don't need hooks, caching and other features (and I think we don't) plain fetch + graphql types are all we need for just a few queries we have here.

Reply:
Okay, so lets start with your solution and we can think later if we can hide it later